// xstream-from-pull-stream/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var xstream_1 = require("xstream");
function xsFromPullStream(pullStream) {
    var producer = {
        isRunning: false,
        start: function (listener) {
            var drain = function drain(read) {
                read(null, function more(end, data) {
                    if (end === true) {
                        listener.complete();
                        return;
                    }
                    if (end) {
                        listener.error(end);
                        return;
                    }
                    listener.next(data);
                    if (producer.isRunning) {
                        read(null, more);
                    }
                });
            };
            producer.isRunning = true;
            try {
                drain(pullStream);
            }
            catch (e) {
                listener.error(e);
            }
        },
        stop: function () {
            producer.isRunning = false;
        }
    };
    return xstream_1.default.create(producer);
}
exports.default = xsFromPullStream;
