// cycle-native-navigation/lib/NavSource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var xstream_1 = require("xstream");
var react_native_navigation_1 = require("react-native-navigation");
var NavSource = /** @class */ (function () {
    function NavSource() {
        this._topBar = xstream_1.default.create();
        this._back = xstream_1.default.create();
        this._didAppear = xstream_1.default.create();
        this._didDisappear = xstream_1.default.create();
        this._globalDidDisappear = xstream_1.default.create({
            start: function (listener) {
                react_native_navigation_1.Navigation.events().registerComponentDidDisappearListener(function (event) {
                    listener.next(event);
                });
            },
            stop: function () { },
        });
    }
    NavSource.prototype.topBarButtonPress = function (buttonId) {
        if (buttonId)
            return this._topBar.filter(function (id) { return id === buttonId; });
        else
            return this._topBar;
    };
    NavSource.prototype.backPress = function () {
        return this._back;
    };
    NavSource.prototype.didAppear = function () {
        return this._didAppear;
    };
    NavSource.prototype.didDisappear = function () {
        return this._didAppear;
    };
    NavSource.prototype.globalDidDisappear = function (componentName) {
        if (componentName) {
            return this._globalDidDisappear.filter(function (ev) { return ev.componentName === componentName; });
        }
        else {
            return this._globalDidDisappear;
        }
    };
    return NavSource;
}());
exports.NavSource = NavSource;
