// @cycle/react/lib/cjs/scope.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var xstream_1 = require("xstream");
var Scope = /** @class */ (function () {
    function Scope() {
        this.handlers = {};
        this.listeners = {};
    }
    Scope.prototype.getSelectorHandlers = function (selector) {
        return this.handlers[selector] || {};
    };
    Scope.prototype.getHandler = function (selector, evType) {
        var sel = selector;
        this.handlers[sel] = this.handlers[sel] || {};
        if (!this.handlers[sel][evType]) {
            this.handlers[sel][evType] = xstream_1.default.create();
            if (this.listeners[sel]) {
                this.listeners[sel]();
            }
        }
        return this.handlers[sel][evType];
    };
    Scope.prototype.subscribe = function (selector, listener) {
        var sel = selector;
        this.listeners[sel] = listener;
        var that = this;
        return function unsubscribe() {
            delete that.listeners[sel];
        };
    };
    return Scope;
}());
exports.Scope = Scope;
