// cycle-native-toastandroid/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var react_native_1 = require("react-native");
exports.Duration = {
    SHORT: react_native_1.ToastAndroid.SHORT,
    LONG: react_native_1.ToastAndroid.LONG,
};
exports.Gravity = {
    TOP: react_native_1.ToastAndroid.TOP,
    CENTER: react_native_1.ToastAndroid.CENTER,
    BOTTOM: react_native_1.ToastAndroid.BOTTOM,
};
function makeToastDriver() {
    return function toastDriver(sink) {
        if (react_native_1.Platform.OS === 'android') {
            sink.addListener({
                next: function (t) {
                    var args = [
                        t.message,
                        t.duration,
                        t.gravity,
                    ].filter(function (x) { return typeof x !== 'undefined'; });
                    react_native_1.ToastAndroid[t.type].apply(react_native_1.ToastAndroid, args);
                },
            });
        }
    };
}
exports.makeToastDriver = makeToastDriver;
