// cycle-native-keyboard/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var adapt_1 = require("@cycle/run/lib/adapt");
var xstream_1 = require("xstream");
var react_native_1 = require("react-native");
function makeStream(type) {
    return xstream_1.default.create({
        start: function (listener) {
            this.keyboardCallback = function (ev) {
                listener.next(ev);
            };
            react_native_1.Keyboard.addListener(type, this.keyboardCallback);
        },
        stop: function () {
            react_native_1.Keyboard.removeListener(type, this.keyboardCallback);
        },
    });
}
var KeyboardSource = /** @class */ (function () {
    function KeyboardSource() {
        this._map = new Map();
        var types = [
            'keyboardWillShow',
            'keyboardDidShow',
            'keyboardWillHide',
            'keyboardDidHide',
            'keyboardWillChangeFrame',
            'keyboardDidChangeFrame',
        ];
        for (var _i = 0, types_1 = types; _i < types_1.length; _i++) {
            var type = types_1[_i];
            this._map.set(type, makeStream(type));
        }
    }
    KeyboardSource.prototype.events = function (type) {
        return adapt_1.adapt(this._map.get(type));
    };
    return KeyboardSource;
}());
exports.KeyboardSource = KeyboardSource;
function makeKeyboardDriver() {
    return function keyboardDriver(sink) {
        sink.addListener({
            next: function (t) {
                react_native_1.Keyboard.dismiss();
            },
        });
        return new KeyboardSource();
    };
}
exports.makeKeyboardDriver = makeKeyboardDriver;
