// ssb-typescript/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function isMsg(msg) {
    return msg && msg.key && msg.value && typeof msg.value === 'object';
}
exports.isMsg = isMsg;
function hasContent(msg) {
    return !!(msg && msg.value && msg.value.content);
}
exports.hasContent = hasContent;
function isPostMsg(msg) {
    return hasContent(msg) && msg.value.content.type === 'post';
}
exports.isPostMsg = isPostMsg;
function isRootPostMsg(msg) {
    return isPostMsg(msg) && !msg.value.content.root;
}
exports.isRootPostMsg = isRootPostMsg;
function isReplyPostMsg(msg) {
    return isPostMsg(msg) && !!msg.value.content.root;
}
exports.isReplyPostMsg = isReplyPostMsg;
function isAboutMsg(msg) {
    return hasContent(msg) && msg.value.content.type === 'about';
}
exports.isAboutMsg = isAboutMsg;
function isContactMsg(msg) {
    return hasContent(msg) && msg.value.content.type === 'contact';
}
exports.isContactMsg = isContactMsg;
function isVoteMsg(msg) {
    return hasContent(msg) && msg.value.content.type === 'vote';
}
exports.isVoteMsg = isVoteMsg;
function isPrivate(msg) {
    return hasContent(msg) && typeof msg.value.content === 'string';
}
exports.isPrivate = isPrivate;
function isPublic(msg) {
    return hasContent(msg) && typeof msg.value.content === 'object';
}
exports.isPublic = isPublic;
